src/dataflow-types/src/client.rs

```diff
@@ -658,6 +667,120 @@ pub mod partitioned {
+            let mut offset = 0;
+            let mut index = 1;
+            while index < self.buffer.len()
+                && PartialOrder::less_than(&Antichain::from_elem(self.buffer[index].1), &upper)
+            {
+                if self.buffer[index].0 == self.buffer[offset].0
+                    && self.buffer[index].1 == self.buffer[offset].1
+                {
+                    self.buffer[offset].2 += self.buffer[index].2;
+                } else {
+                    if self.buffer[offset].2 != 0 {
+                        offset += 1;
+                    }
+                    let next = std::mem::take(&mut self.buffer[index]);
+                    self.buffer[offset] = next;
+                }
+                index += 1;
+            }
+            if offset < self.buffer.len() && self.buffer[offset].2 != 0 {
+                offset += 1;
+            }
+            self.buffer.drain(0..offset).collect()
+        }
+
+        pub fn push_data<I: IntoIterator<Item = (Row, Timestamp, Diff)>>(&mut self, data: I) {
+            self.buffer.extend(data);
+        }
+
+        pub fn record_progress(
+            &mut self,
+            shard_id: usize,
+            progress: Antichain<Timestamp>,
+        ) -> Option<Antichain<Timestamp>> {
+            // In the future, we can probably replace all this logic with the use of a `MutableAntichain`.
+            // We would need to have the tail workers report both their old frontier and their new frontier, rather
+            // than just the latter. But this will all be subsumed anyway by the proposed refactor to make TAILs
+            // behave like PEEKs.
+            match self.frontiers.entry(shard_id) {
+                Entry::Occupied(mut oe) => {
+                    assert!(
+                        PartialOrder::less_than(oe.get(), &progress),
+                        "Timestamps should advance"
+                    );
+                    oe.insert(progress);
+                }
+                Entry::Vacant(ve) => {
+                    ve.insert(progress);
+                }
+            }
+            assert!(self.frontiers.len() <= self.parts);
+            if self.frontiers.len() == self.parts {
+                let min_frontier = self
+                    .frontiers
+                    .values()
+                    .min_by(|f1, f2| {
+                        if PartialOrder::less_than(*f1, *f2) {
+                            Ordering::Less
+                        } else if f1 == f2 {
+                            Ordering::Equal
+                        } else if PartialOrder::less_than(*f2, *f1) {
+                            Ordering::Greater
+                        } else {
+                            panic!("Tail requires totally-ordered timestamps!")
+                        }
+                    })
+                    .unwrap()
+                    .clone();
+                // assert!(self.reported_frontier.less_equal(&min_frontier));
```

Comment:
Stray, or important?

Reply:
Oh I see; it is documentation for the next line. nvm!

Reply:
Actually, it was stray, but if it seems like useful documentation, I can leave it 🤷‍♂️